### PR TITLE
Kiosk state machine

### PR DESCRIPTION
**Pull request by podhrmic**
Creates a state machine reflecting the kiosk slides, and describing the actions needed to happen at each state (important for the Python backend).

Rendering here:
state_machine_hacker_kiosk

### CODE REVIEW

File: besspin/cyberPhys/specs/state_machine_hacker_kiosk.txt [hunk 1]
```diff
@@ -1,95 +1,90 @@
...
-'HackerKioskUp: connect to the network
-'HackerKioskUp: start Kiosk UI app
-'HackerKioskReady: wait for user
-'[*] --> HackerKioskUp: Hacker Kiosk powered up
-'HackerKioskUp -[#blue]-> HackerKioskReady: Success\nSend CMD_COMPONENT_READY(HACKER_KIOSK)
-'@enduml
+Slide16 --> Slide16: A hack is selected
+Slide16: All hacks fail with an error message (OTA hack not succesfull)
+Slide16: TODO: add a slide where we attempt to hack OTA server
+Slide16: TODO: OTA server crashes on SSITH P2 when a hack is attempted (~2min downtime)
```
Comment (dmzimmerman) [inline besspin/cyberPhys/specs/state_machine_hacker_kiosk.txt:63]: Is there really a 2 minute downtime when a hack is attempted? Shouldn't the OTA server just crash and then be restarted? Or is that the maximum timeout that would occur before the watchdog restarts it? (presumably we could also just make `service` restart it automatically if it dies...)

File: besspin/cyberPhys/specs/state_machine_hacker_kiosk.txt [hunk 2]
```diff
@@ -1,95 +1,90 @@
...
-    ScenarioBaselineIntro --> ScenarioBaselineExplainOtaHack: Hacker clicks though the intro slides
-    ScenarioBaselineExplainOtaHack --> ScenarioBaselineHackedOTA: Hacker clicks in "Hack OTA update server"\nSend CMD_LED_SEQ(OTA_HACK)
-    state "OTA server hacked" as ScenarioBaselineHackedOTA
-    ScenarioBaselineHackedOTA: signing key erased
+state "Slide 6a HACK OTA SERVER" as Slide6_A
+Slide6_A: Call "OtaHack.hackServer()"
 
-    state "Hack infotainment" as ScenarioBaselineHackInfotainment
-    ScenarioBaselineHackInfotainment: show available infotainment hacks
+Slide6_A --> Slide6_B: "OtaHack.hackServer()" returned True
```
Comment (dmzimmerman) [inline besspin/cyberPhys/specs/state_machine_hacker_kiosk.txt:23]: What happens if it didn't return `True`? Do we have (or need) an escape hatch if the hack doesn't work/something goes wrong?

File: besspin/cyberPhys/specs/state_machine_hacker_kiosk.txt [hunk 3]
```diff
@@ -1,95 +1,90 @@
...
 
-    state "Hack Critical Systems" as ScenarioBaselineHackCriticalSystems
-    ScenarioBaselineHackCriticalSystems --> ScenarioBaselineEndOfScenario: Hacker: Clicks on next slide
-    state "Scenario Done" as ScenarioBaselineEndOfScenario
-}
+state "Slide 10 HACK CRITICAL SYSTEMS" as Slide10
+Slide10: When a hack is selected, an appropriate pre-compiled binary is selected
+Slide10: and uploaded using "OtaHack.uploadFile(binary)"
+Slide10: The hack button will color red/green based on the selection
+Slide10: and whether the upload was succesfull
```
Comment (dmzimmerman) [inline besspin/cyberPhys/specs/state_machine_hacker_kiosk.txt:47]: s/succesfull/successful/

File: besspin/cyberPhys/specs/state_machine_hacker_kiosk.txt [hunk 4]
```diff
@@ -1,95 +1,90 @@
...
-' TODO: handle error conditions
-KioskWait --> KioskReady: Success\nSend CMD_COMPONENT_READY(KIOSK)\nSend CMD_LED_SEQ(BOOT)
-state "Hacker kiosk ready" as KioskReady
-KioskReady: welcome page
 
+Slide3 --> Slide4: Click "Hack the car" button
 
-state "Baseline Scenario" as ScenarioBaseline {
-    state "Scenario intro" as ScenarioBaselineIntro
-    ' ScenarioBaselineIntro: Driver: get faimilar with the controls
```
Comment (dmzimmerman) [inline besspin/cyberPhys/specs/state_machine_hacker_kiosk.txt:24]: s/faimilar/familiar/

File: besspin/cyberPhys/specs/state_machine_hacker_kiosk.txt [hunk 5]
```diff
@@ -1,95 +1,90 @@
...
-'state "Hacker Kiosk Ready" as HackerKioskReady
-'HackerKioskUp: boot OS
-'HackerKioskUp: connect to the network
-'HackerKioskUp: start Kiosk UI app
-'HackerKioskReady: wait for user
-'[*] --> HackerKioskUp: Hacker Kiosk powered up
-'HackerKioskUp -[#blue]-> HackerKioskReady: Success\nSend CMD_COMPONENT_READY(HACKER_KIOSK)
-'@enduml
+Slide16 --> Slide16: A hack is selected
+Slide16: All hacks fail with an error message (OTA hack not succesfull)
```
Comment (dmzimmerman) [inline besspin/cyberPhys/specs/state_machine_hacker_kiosk.txt:61]: s/succesfull/successful/

File: besspin/cyberPhys/specs/state_machine_hacker_kiosk.txt [hunk 6]
```diff
@@ -1,95 +1,90 @@
...
+Slide17 --> Slide20: "Next" button pressed
 
-@enduml
\ No newline at end of file
+state "Slide 20 Everything is hackable" as Slide20
+state "Slide 21 SSITH is The Solution" as Slide21
+Slide20 --> Slide21: "Restart" button pressed
+Slide21 -> Slide3
+
+Slide15 --> Slide18: click "Hack Cricial systems"
```
Comment (dmzimmerman) [inline besspin/cyberPhys/specs/state_machine_hacker_kiosk.txt:75]: s/Cricial/Critical/

File: besspin/cyberPhys/specs/state_machine_hacker_kiosk.txt [hunk 7]
```diff
@@ -1,95 +1,90 @@
...
+Slide18: Tx CMD_ACTIVE_SCENARIO(SCENARIO_SECURE_ECU)
+
+state "Slide 18 SECURE ECU" as Slide18
+
+Slide18 --> Slide18: A hack is selected
+Slide18: When a hack is selected, an appropriate pre-compiled binary is selected
+Slide18: and uploaded using "OtaHack.uploadFile(binary)"
+Slide18: The hack will fail, and lead to short (200ms?) unavailibility of the ECU
```
Comment (dmzimmerman) [inline besspin/cyberPhys/specs/state_machine_hacker_kiosk.txt:83]: s/unavailibility/unavailability/

**Review by dmzimmerman — CHANGES_REQUESTED** [review]
Some very minor typos and a couple of questions, but otherwise this looks good.

File: besspin/cyberPhys/specs/state_machine_hacker_kiosk.txt (hunk 1 above)
Comment (podhrmic) [reply besspin/cyberPhys/specs/state_machine_hacker_kiosk.txt:63]: I measured how long the automatic restart takes a couple of weeks back, and it was around 2 minutes with the auto start (`Restart=on-failure` in the service config). I will re-measure to see how long `systemctl stop/start` takes.

File: besspin/cyberPhys/specs/state_machine_hacker_kiosk.txt (hunk 2 above)
Comment (podhrmic) [reply besspin/cyberPhys/specs/state_machine_hacker_kiosk.txt:23]: Great question. The UI doesn't really have a way of showing failures - so we can at best log it/show it in the console. What do you think?

Comment (dmzimmerman) [reply besspin/cyberPhys/specs/state_machine_hacker_kiosk.txt:23]: I think that's fine as far as it goes... but it's not clear what the UI should _do_ in the event of a failure. We can't really just move on to the infotainment hack screen.

**Comment by lolsborn** [comment]
For now, I could just rely on triggering a javascript alert() which is a modal dialog.

On Wed, Jun 9, 2021, at 1:24 PM, Daniel M. Zimmerman wrote:
> 

> ***@***.**** commented on this pull request.

> 
> In besspin/cyberPhys/specs/state_machine_hacker_kiosk.txt <https://github.com/GaloisInc/BESSPIN-Tool-Suite/pull/1192#discussion_r648659225>:

> >  
-    state "Hack infotainment" as ScenarioBaselineHackInfotainment
-    ScenarioBaselineHackInfotainment: show available infotainment hacks
+Slide6_A --> Slide6_B: "OtaHack.hackServer()" returned True

**Comment by dmzimmerman** [comment]
> For now, I could just rely on triggering a javascript alert() which is a modal dialog.

Sounds reasonable.

**Review by dmzimmerman — APPROVED** [review]
I think you need to regenerate the PNG with the spelling fixes, but otherwise this is fine.

**Pull request merged**